backend/census_historical_migration/workbooklib/post_upload_utils.py

```diff
@@ -0,0 +1,162 @@
+    extract_additional_ueis,
+    extract_additional_eins,
+)
+import audit.validators
+
+from audit.fixtures.excel import FORM_SECTIONS
+
+logger = logging.getLogger(__name__)
+
+
+def get_field_by_section(sac, section):
+    if section == FORM_SECTIONS.FEDERAL_AWARDS_EXPENDED:
+        return sac.federal_awards
+    elif section == FORM_SECTIONS.FINDINGS_UNIFORM_GUIDANCE:
+        return sac.findings_uniform_guidance
+    elif section == FORM_SECTIONS.FINDINGS_TEXT:
+        return sac.findings_text
+    elif section == FORM_SECTIONS.CORRECTIVE_ACTION_PLAN:
+        return sac.corrective_action_plan
+    elif section == FORM_SECTIONS.SECONDARY_AUDITORS:
+        return sac.secondary_auditors
+    elif section == FORM_SECTIONS.NOTES_TO_SEFA:
+        return sac.notes_to_sefa
+    elif section == FORM_SECTIONS.ADDITIONAL_UEIS:
+        return sac.additional_ueis
+
+
+extract_mapping = {
+    FORM_SECTIONS.FEDERAL_AWARDS_EXPENDED: extract_federal_awards,
+    FORM_SECTIONS.FINDINGS_UNIFORM_GUIDANCE: extract_findings_uniform_guidance,
+    FORM_SECTIONS.FINDINGS_TEXT: extract_findings_text,
+    FORM_SECTIONS.CORRECTIVE_ACTION_PLAN: extract_corrective_action_plan,
+    FORM_SECTIONS.SECONDARY_AUDITORS: extract_secondary_auditors,
+    FORM_SECTIONS.NOTES_TO_SEFA: extract_notes_to_sefa,
+    FORM_SECTIONS.ADDITIONAL_UEIS: extract_additional_ueis,
+    FORM_SECTIONS.ADDITIONAL_EINS: extract_additional_eins,
+}
+
+validator_mapping = {
+    FORM_SECTIONS.FEDERAL_AWARDS_EXPENDED: audit.validators.validate_federal_award_json,
+    FORM_SECTIONS.FINDINGS_UNIFORM_GUIDANCE: audit.validators.validate_findings_uniform_guidance_json,
+    FORM_SECTIONS.FINDINGS_TEXT: audit.validators.validate_findings_text_json,
+    FORM_SECTIONS.CORRECTIVE_ACTION_PLAN: audit.validators.validate_corrective_action_plan_json,
+    FORM_SECTIONS.SECONDARY_AUDITORS: audit.validators.validate_secondary_auditors_json,
+    FORM_SECTIONS.NOTES_TO_SEFA: audit.validators.validate_notes_to_sefa_json,
+    FORM_SECTIONS.ADDITIONAL_UEIS: audit.validators.validate_additional_ueis_json,
+    FORM_SECTIONS.ADDITIONAL_EINS: audit.validators.validate_additional_eins_json,
+    "PDF": audit.validators.validate_single_audit_report_file,
+}
+
+
+def _post_upload_pdf(this_sac, this_user, pdf_filename):
+    """Upload a workbook for this SAC.
+
+    This should be idempotent if it is called on a SAC that already
+    has a federal awards file uploaded.
+    """
+    PDFFile = apps.get_model("audit.SingleAuditReportFile")
+
+    if PDFFile.objects.filter(sac_id=this_sac.id).exists():
+        # there is already an uploaded file and data in the object so
+        # nothing to do here
+        return
+
+    with open(pdf_filename, "rb") as f:
+        content = f.read()
+    file = SimpleUploadedFile(pdf_filename, content, "application/pdf")
+    print(file.__dict__)
+    pdf_file = PDFFile(
+        file=file,
+        component_page_numbers={  # FIXME - How do we want to handle these values?
```

Comment:
They might actually exist somewhere in the historic data. For now, I don't know. We should start a ticket of questions like this. Or something.

Reply:
https://github.com/GSA-TTS/FAC/issues/2832